settings-portal/Settings.vala

```diff
@@ -0,0 +1,156 @@
+            );
+
+            user_path = accounts_service.find_user_by_name (GLib.Environment.get_user_name ());
+        } catch (Error e) {
+            critical (e.message);
+        }
+    }
+
+    private void setup_prefers_color_scheme () {
+        try {
+            pantheon_act = GLib.Bus.get_proxy_sync (
+                GLib.BusType.SYSTEM,
+                "org.freedesktop.Accounts",
+                user_path,
+                GLib.DBusProxyFlags.GET_INVALIDATED_PROPERTIES
+            );
+
+            color_scheme = pantheon_act.prefers_color_scheme;
+
+            ((GLib.DBusProxy) pantheon_act).g_properties_changed.connect ((changed, invalid) => {
+                var value = changed.lookup_value ("PrefersColorScheme", new VariantType ("i"));
+                if (value != null) {
+                    color_scheme = value.get_int32 ();
+                }
+            });
+        } catch (Error e) {
+            critical (e.message);
+        }
+    }
+}
+
+[DBus (name = "org.freedesktop.impl.portal.Settings")]
+public class SettingsDaemon.Settings : GLib.Object {
+    private GLib.DBusConnection connection;
+
+    public uint32 version {
+        get { return 1; }
+    }
+
+    public signal void setting_changed (string namespace, string key, GLib.Variant value);
+
+    private AccountsServiceMonitor monitor;
+
+    public Settings (GLib.DBusConnection connection) {
+        this.connection = connection;
+    }
+
+    construct {
+        monitor = new AccountsServiceMonitor ();
+        monitor.notify["color-scheme"].connect (() => {
+            var color_scheme = new Variant.uint32 (monitor.color_scheme);
+            setting_changed ("org.freedesktop.appearance", "color-scheme", get_color_scheme ());
+        });
+    }
+
+    private bool namespace_matches (string namespace, string[] patterns) {
+        foreach (var pattern in patterns) {
+            if (pattern[0] == '\0') {
+                return true;
+            }
+
+            if (pattern == namespace) {
+                return true;
+            }
+
+            int pattern_len = pattern.length;
+            if (pattern[pattern_len - 1] == '*' && Posix.strncmp (namespace, pattern, pattern_len - 1) == 0) {
+                return true;
+            }
+        }
+
+        return patterns.length == 0;
+    }
+
+    private GLib.Variant get_color_scheme () {
+        return new GLib.Variant.uint32 (monitor.color_scheme);
+    }
+
+    public async void read_all (string[] namespaces, out GLib.Variant value) throws GLib.DBusError, GLib.IOError {
+        var builder = new GLib.VariantBuilder (new GLib.VariantType ("(a{sa{sv}})"));
+
+        builder.open (new GLib.VariantType ("a{sa{sv}}"));
+
+        if (namespace_matches ("org.freedesktop.appearance", namespaces)) {
+            var dict = new VariantDict ();
+
+            dict.insert_value ("color-scheme", get_color_scheme ());
+
+            builder.add ("{sa{sv}}", "org.freedesktop.appearance", dict.end ());
```

Comment:
It crashes here, not sure offhand why.

```vala
alexm@Standard-PC-Q35-ICH9-2009-32680571:~/Projects/settings-daemon$ /usr/libexec/io.elementary.settings-daemon.xdg-desktop-portal  -r

(io.elementary.settings-daemon.xdg-desktop-portal:41074): GLib-CRITICAL **: 21:47:53.348: g_variant_builder_end: assertion 'ensure_valid_builder (builder)' failed

(io.elementary.settings-daemon.xdg-desktop-portal:41074): GLib-CRITICAL **: 21:47:53.348: g_variant_get_type: assertion 'value != NULL' failed

(io.elementary.settings-daemon.xdg-desktop-portal:41074): GLib-CRITICAL **: 21:47:53.348: g_variant_type_is_array: assertion 'g_variant_type_check (type)' failed

(io.elementary.settings-daemon.xdg-desktop-portal:41074): GLib-CRITICAL **: 21:47:53.348: g_variant_get_type_string: assertion 'value != NULL' failed

(io.elementary.settings-daemon.xdg-desktop-portal:41074): GLib-ERROR **: 21:47:53.348: g_variant_new: expected array GVariantBuilder but the built value has type '(null)'
Trace/breakpoint trap
```

Reply:
As suggested by Marukesu, stopped using `VariantBuilder` there completely so this is fixed.